### ISSUE TITLE
Shims ICE when a scalar argument isn't actually a scalar

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#3842]
For instance, consider:
```rust
extern "C" {
    fn pipe(pipefd: [i32; 2]) -> i32;
}

fn main() {
    let mut fds: [i32; 2] = [0; 2]; 
    assert_eq!(unsafe { pipe(fds) }, 0); 
}
```
I think to fix this properly we need to completely re-do the way we handle shim arguments: every `OpTy` needs to be transmuted to the right type before we do anything else with it. (And we can't even just call `OpTy::transmute` as that will ICE if the transmute would change the size. Though maybe we should change that.)

We probably want a macro that helps with that.

**Comment by RalfJung** [comment]
I think we want a new function like `get_arg` or so that preprocesses the `OpTy`. Given an `op: OpTy` and `ty: Ty`, we do a check like
- `let layout = layout_of(ty);`
- Check `layout.abi.eq_up_to_validity(&op.layout.abi)`. If that is `false`, the argument type is invalid -> raise UB.
- Return `op.transmute(layout)`.

**Comment by RalfJung** [comment]
Similar ICEs can occur when the return type of the function is wrong -- we end up calling `write_scalar` with a scalar of the wrong size. This either fails to initialize a part of the result (if the scalar is smaller than the return place), or ICEs since we are exceeding the bounds of our `AllocRefMut` (if the scalar is bigger than the return place).

**Comment by geetanshjuneja** [comment]
Can I look into this task?

**Comment by RalfJung** [comment]
Definitely! However, this will require some design work. As mentioned in the issue description, we don't have a clear plan yet for how to best fix this issue.

**Comment by geetanshjuneja** [comment]
what does this fn `eq_up_to_validity` do?  I couldn't find its definition anywhere?

**Comment by tiif** [comment]
Quick answer: It is part of the rustc https://github.com/rust-lang/rust/blob/64e06c0f5578829373743884b708d494136c3e8f/compiler/rustc_abi/src/lib.rs#L1517.

**Comment by RalfJung** [comment]
Please use [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/269128-miri) for basic questions about how Miri works; the issue should be for discussion specific to the task described above. :)

**Comment by RalfJung** [comment]
With https://github.com/rust-lang/miri/pull/4185 having landed, there's a whole lot of low-hanging fruit here that is well-suited for first contributors: pick a family of related shim implementations, and convert them from `check_shim` to `check_shim_abi`. For now, skip any shims that take or return any non-trivial types (e.g. `struct`s or really anything except for integers and pointers). We have exactly one example of using `check_shim_abi` that you can follow:

https://github.com/rust-lang/miri/blob/74f4d67ae9e64b04512a6d52ec9775429c379bd9/src/shims/unix/foreign_items.rs#L203-L210

The main work here is figuring out the argument and return types for all those functions.

**Comment by geetanshjuneja** [comment]
Hi, sorry I forgot about this. I'll start working on this asap.

**Comment by RalfJung** [comment]
Seeing how things look like in https://github.com/rust-lang/miri/pull/4263, I wonder if it's worth playing around a bit with a macro to make these expected signatures easier to read and write. I am imagining that instead of 
```rust
                let [name, value, overwrite] = this.check_shim_abi(
                    link_name,
                    abi,
                    ExternAbi::C { unwind: false },
                    [
                        this.machine.layouts.const_raw_ptr.ty,
                        this.machine.layouts.const_raw_ptr.ty,
                        this.tcx.types.i32,
                    ],
                    this.tcx.types.i32,
                    args,
                )?;
```
We could have something like
```rust
                let [name, value, overwrite] = this.check_shim_abi(
                    link_name,
                    abi,
                    ExternAbi::C { unwind: false },
                    shim_sig!(*const _, *const _, i32 -> i32),
                    args,
                )?;
```
This should probably be done before we port the rest of Miri, or else all `check_shim` calls will have to be adjusted again.

We could even include the ABI, making it something like
```rust
                let [name, value, overwrite] = this.check_shim_abi(
                    link_name,
                    abi,
                    shim_sig!(extern "C" fn(*const _, *const _, i32) -> i32),
                    args,
                )?;
```
I think I quite like the idea of using function pointer syntax for this.

At that point I'd advocate for reordering the arguments a bit as well, making this
```rust
                let [name, value, overwrite] = this.check_shim_abi(
                    shim_sig!(extern "C" fn(*const _, *const _, i32) -> i32),
                    link_name,
                    abi,
                    args,
                )?;
```

**Comment by geetanshjuneja** [comment]
Hi, I can work on implementing this macro.

**Comment by RalfJung** [comment]
Yeah, feel free to give it a shot. :)
